From 16242ce1f1d006f20317a2fa9baab2debf39bd46 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Thu, 8 May 2025 14:18:45 +0800
Subject: [PATCH 84/95] mtk: mt76: mt7996: fix waning when removing erroneous
 links

This commit fixes the following warning:
[  279.034200] ------------[ cut here ]------------
[  279.038807] destroying interface with valid links 0x0003
[  279.044141] WARNING: CPU: 3 PID: 2559 at ieee80211_do_stop+0x94c/0x990 [mac80211]
[  279.051672] Modules linked in: ksmbd pppoe ppp_async nft_fib_inet nf_flow_table_inet l2tp_ppp pptp pppox ppp_mppe ppp_generic nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nfte
[  279.051816]  iptable_filter ipt_ah ipt_ECN ip_tables compat(O) br_netfilter at24 mt76qos(O) crypto_safexcel ntfs3 sg pwm_fan i2c_gpio i2c_algo_bit i2c_mux_pca954x i2c_mux ip6tab]
[  279.215572] CPU: 3 PID: 2559 Comm: hostapd Tainted: G           O       6.6.86 #0
[  279.223042] Hardware name: MediaTek MT7988A Reference Board (DT)
[  279.229036] pstate: 80400005 (Nzcv daif +PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[  279.235985] pc : ieee80211_do_stop+0x94c/0x990 [mac80211]
[  279.241411] lr : ieee80211_do_stop+0x94c/0x990 [mac80211]
[  279.246833] sp : ffffffc085c53a60
[  279.250135] x29: ffffffc085c53a60 x28: ffffff8000958d00 x27: 0000000000000001
[  279.257259] x26: ffffffc0797c1694 x25: ffffff8017bb4000 x24: 0000000000000001
[  279.264383] x23: 0000000000001003 x22: 0000000000000000 x21: ffffffc0797c1694
[  279.271506] x20: ffffff80059611e0 x19: ffffff8017bb4900 x18: 000000000000098f
[  279.278630] x17: ffffffbfbf023000 x16: ffffffc080f90000 x15: ffffffc080db5948
[  279.285754] x14: 0000000000001cad x13: 000000000000098f x12: 00000000ffffffea
[  279.292876] x11: 00000000ffffefff x10: ffffffc080e0d948 x9 : ffffffc080db58f0
[  279.299999] x8 : 0000000000017fe8 x7 : c0000000ffffefff x6 : 0000000000000001
[  279.307121] x5 : ffffff803fd9b8d0 x4 : 0000000000000000 x3 : 0000000000000027
[  279.314244] x2 : 0000000000000027 x1 : 0000000000000023 x0 : 000000000000002c
[  279.321368] Call trace:
[  279.323804]  ieee80211_do_stop+0x94c/0x990 [mac80211]
[  279.328880]  ieee80211_stop+0x60/0x194 [mac80211]
[  279.333610]  __dev_close_many+0xa0/0x104
[  279.337525]  __dev_change_flags+0xf0/0x210
[  279.341611]  dev_change_flags+0x20/0x64

This warning usually happens when some of links setup failed.
The root cause is that the erroneous links are not finishing setting in
mt76 yet (e.g. struct mt7996_vif_link not ready), but mcu commands like
mt7996_mcu_mld_reconf_stop_link() are still sending during link remove
process, which causes FW to return error code and thus
mt7996_change_vif_links() returns failed to upper layer, making links
unremoved.

To fix this issue, do not send commands for those erroneous links which
haven't finished setting up in mt76.

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 mt7996/main.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index d67b970e..959725e5 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -2784,7 +2784,8 @@ mt7996_change_vif_links(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
 	struct mt7996_phy *phy = mvif->deflink.phy;
 	struct mt7996_vif_link *mconf;
-	unsigned long rem = old_links & ~new_links & ~vif->dormant_links;
+	unsigned long rem = old_links & ~new_links & ~vif->dormant_links &
+			    mvif->mt76.valid_links;
 	unsigned long add = new_links & ~old_links;
 	int link_id, ret = 0;
 
-- 
2.45.2

